staging/src/k8s.io/apiserver/pkg/endpoints/openapi/restricted_source.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+// SchemaCacher fetches the openapi schema once and then caches it in memory
+type restrictedSchemaSource struct {
+	source       SchemaSource
+	allowedGroup string
+}
+
+var _ SchemaSource = &restrictedSchemaSource{}
+
+// NewSchemaSourceForGroup wraps SchemaSource so Resources objects returned by it
+// will only allow lookups within a certain group.
+func NewSchemaSourceForGroup(source SchemaSource, allowedGroup string) SchemaSource {
```

Comment:
If types are only allowed to reference within a GroupVersion, then this should be changed to
```func NewSchemaSourceForGroupVersion(source SchemaSource, allowedGroupVersion schema.GroupVersion) SchemaSource {```